Allow to read openstack cloud provider config from a secret

Fedosin opened the pull request:
This PR fixes openstack cloudprovider to read secrets from informer, so as it is consistent with other cloudproviders that need to
read kubernetes objects.

/kind bug

```release-note
Fix openstack cloudprovider to read secrets from informer, so as it is consistent with other cloudproviders that need to
read kubernetes objects
```

sttts commented:
Sgtm as a fix-up of an existing carry. Please sync with next rebase team how to handle the upstream PR rename.

gnufied commented:
afaict - this PR attempts fixing two things (mainly):

1. It removes use of kubeconfig for reading secrets and uses informer. Openstack cloudprovider already supported reading cloud-config from secrets, so this change is not a new feature I think.
2. The second fix is - it uses metadata service to fetch instance related information in kubelet. Does this mean - we don't need cloud-provider configuration in kubelet anymore? Can we mark all functions within cloudprovider that is called by kubelet?

Fedosin committed:
Use openstack metadata service in kubelet

Instead of sending requests to nova kubelet can fetch
node info from metadata service. It allows to skip
cloud provider initialization when credentials are
stored in a secret.

Fedosin committed:
Fix openstack cloud provider to read secret from informer

Currently reading secret data is broken in the cloud
provider. This commit fixes it by starting using
informers.

Co-authored-by: Hemant Kumar <hekumar@redhat.com>
Co-authored-by: Jan Safranek <jsafrane@redhat.com>

Fedosin commented:
/retest

Fedosin commented:
/retest

Fedosin commented:
/kind bug

mdbooth commented on plugin/pkg/admission/storage/persistentvolume/label/admission.go:
```diff
@@ -175,7 +193,7 @@ func (l *persistentVolumeLabel) findVolumeLabels(volume *api.PersistentVolume) (
 	domain, domainOK := existingLabels[v1.LabelTopologyZone]
 	region, regionOK := existingLabels[v1.LabelTopologyRegion]
-	// If they don't have GA labels we should check for failuredomain beta labels
+	// If they dont have GA labels we should check for failuredomain beta labels
```
Comment damage.

mdbooth commented on plugin/pkg/admission/storage/persistentvolume/label/admission.go:
```diff
@@ -85,6 +90,19 @@ func newPersistentVolumeLabel() *persistentVolumeLabel {
 func (l *persistentVolumeLabel) SetCloudConfig(cloudConfig []byte) {
 	l.cloudConfig = cloudConfig
 }
```
Missing blank line.

mdbooth reviewed (commented):
It doesn't look like we're watching the cloud secret, so presumably if we update it nothing will notice. Are we ok with that?

gnufied commented on staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:
```diff
@@ -48,6 +48,19 @@ const (
+	// can read all required data from metadata service. Absence of
+	// secretLister means that we use kubelet.
+	if os.secretLister == nil {
+		return &Instances{
```
while testing this change - what I found is, kubelet still needs cloud-config even if it is invalid one. Can we make `cloud-config` an optional parameter for cloud-provider initialization so as kubelet does not crashes if no cloud-config is provided?

Fedosin replied:
Done! now `cloud-provider` is an optional parameter.
